File: sources/modules/libraries/networking-api/src/main/java/com/roxiemobile/networkingapi/network/rest/CallResult.java
package com.roxiemobile.networkingapi.network.rest;

import com.roxiemobile.androidcommons.diagnostics.Guard;
import com.roxiemobile.networkingapi.network.rest.response.ResponseEntity;
import com.roxiemobile.networkingapi.network.rest.response.RestApiError;

import org.jetbrains.annotations.NotNull;

public class CallResult<T> extends Result<ResponseEntity<T>, RestApiError, T> {

// MARK: - Construction

    private CallResult(@NotNull ResponseEntity<T> success) {
        super(success);
    }

    private CallResult(@NotNull RestApiError failure) {
        super(failure);
    }

// MARK: - Methods

    public static @NotNull <T> CallResult<T> success(@NotNull ResponseEntity<T> entity) {
        Guard.notNull(entity, "entity is null");
        return new CallResult<>(entity);
    }

    public static @NotNull <T> CallResult<T> failure(@NotNull RestApiError error) {
        Guard.notNull(error, "error is null");
        return new CallResult<>(error);
    }
}

File: sources/modules/libraries/networking-api/src/main/java/com/roxiemobile/networkingapi/network/rest/CallResult.kt
package com.roxiemobile.networkingapi.network.rest

import com.roxiemobile.networkingapi.network.rest.response.ResponseEntity
import com.roxiemobile.networkingapi.network.rest.response.RestApiError

@Deprecated("Must be removed", ReplaceWith("kotlin.Result"))
class CallResult<T>: Result<ResponseEntity<T>, RestApiError, T> {

// MARK: - Construction

    private constructor(responseEntity: ResponseEntity<T>): super(responseEntity)

    private constructor(restApiError: RestApiError): super(restApiError)

// MARK: - Companion

    companion object {

        fun <T> success(responseEntity: ResponseEntity<T>): CallResult<T> {
            return CallResult(responseEntity)
        }

        fun <T> failure(restApiError: RestApiError): CallResult<T> {
            return CallResult(restApiError)
        }
    }
}
